tests/trainer/test_trainer.py

```diff
@@ -1969,3 +1969,49 @@ def training_step(self, batch, batch_idx):
     # simulate random failure in training_step on rank 0
     with pytest.raises(DeadlockDetectedException, match="CustomException"):
         trainer.fit(model)
+
+
+@RunIf(min_gpus=1)
+def test_multiple_trainer_constant_memory_allocated(tmpdir):
+    """
+    This tests ensures calling the trainer several times doesn't increase memory allocated.
+    """
+
+    class TestModel(BoringModel):
+
+        def __init__(self):
+            super().__init__()
+            self._example_input_array = torch.zeros((2, 32))
+
+        @property
+        def example_input_array(self):
+            return self._example_input_array
+
+        def configure_optimizers(self):
+            optimizer = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
+            return [optimizer], [lr_scheduler]
+
+    initial = torch.cuda.memory_allocated(0)
+
+    model = TestModel()
+    trainer_kwargs = dict(default_root_dir=tmpdir, fast_dev_run=True, gpus=1, accelerator="ddp")
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model)
+
+    assert model._example_input_array.device == torch.device("cpu")
+    assert list(trainer.optimizers[0].state.values())[0]["exp_avg_sq"].device == torch.device("cpu")
+    assert trainer.optimizers[0].state
+
+    before = torch.cuda.memory_allocated(0)
+    deepcopy(trainer)
+    after = torch.cuda.memory_allocated(0)
+    torch.cuda.empty_cache()
+    assert before == after == 1024
+
+    trainer_2 = Trainer(**trainer_kwargs)
+    trainer_2.fit(model)
+    after_2 = torch.cuda.memory_allocated(0)
+
+    # todo: (tchaton) Still some memory leaks, could not find the source.
+    assert initial + 2048 == before + 1024 == after_2
```

Comment:
So if there were no more leaks, what should this expression be?

Reply:
initial == before == after_2